File: lua/test/person_entity_test.lua
-- Person entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("tvmaze_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("PersonEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:Person(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = person_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"list", "load"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "person." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set TVMAZE_TEST_PERSON_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- Bootstrap entity data from existing test data.
    local person_ref01_data_raw = vs.items(helpers.to_map(
      vs.getpath(setup.data, "existing.person")))
    local person_ref01_data = nil
    if #person_ref01_data_raw > 0 then
      person_ref01_data = helpers.to_map(person_ref01_data_raw[1][2])
    end

    -- LIST
    local person_ref01_ent = client:Person(nil)
    local person_ref01_match = {}

    local person_ref01_list_result, err = person_ref01_ent:list(person_ref01_match, nil)
    assert.is_nil(err)
    assert.is_table(person_ref01_list_result)

    -- LOAD
    local person_ref01_match_dt0 = {
      id = person_ref01_data["id"],
    }
    local person_ref01_data_dt0_loaded, err = person_ref01_ent:load(person_ref01_match_dt0, nil)
    assert.is_nil(err)
    local person_ref01_data_dt0_load_result = helpers.to_map(person_ref01_data_dt0_loaded)
    assert.is_not_nil(person_ref01_data_dt0_load_result)
    assert.are.equal(person_ref01_data_dt0_load_result["id"], person_ref01_data["id"])

  end)
end)

function person_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/person/PersonTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read person test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "person01", "person02", "person03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("TVMAZE_TEST_PERSON_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["TVMAZE_TEST_PERSON_ENTID"] = idmap,
    ["TVMAZE_TEST_LIVE"] = "FALSE",
    ["TVMAZE_TEST_EXPLAIN"] = "FALSE",
    ["TVMAZE_APIKEY"] = "NONE",
  })

  local idmap_resolved = helpers.to_map(
    env["TVMAZE_TEST_PERSON_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["TVMAZE_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
        apikey = env["TVMAZE_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["TVMAZE_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["TVMAZE_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end

File: go/test/person_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/tvmaze-sdk/go"
	"github.com/voxgig-sdk/tvmaze-sdk/go/core"

	vs "github.com/voxgig-sdk/tvmaze-sdk/go/utility/struct"
)

func TestPersonEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Person(nil)
		if ent == nil {
			t.Fatal("expected non-nil PersonEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"person": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.Person(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.MakeConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.Person(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := personBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"list", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "person." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set TVMAZE_TEST_PERSON_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		personRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.person", setup.data)))
		var personRef01Data map[string]any
		if len(personRef01DataRaw) > 0 {
			personRef01Data = core.ToMapAny(personRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = personRef01Data

		// LIST
		personRef01Ent := client.Person(nil)
		personRef01Match := map[string]any{}

		personRef01ListResult, err := personRef01Ent.List(personRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		_, personRef01ListOk := personRef01ListResult.([]any)
		if !personRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", personRef01ListResult)
		}

		// LOAD
		personRef01MatchDt0 := map[string]any{
			"id": personRef01Data["id"],
		}
		personRef01DataDt0Loaded, err := personRef01Ent.Load(personRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		personRef01DataDt0LoadResult := core.ToMapAny(personRef01DataDt0Loaded)
		if personRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if personRef01DataDt0LoadResult["id"] != personRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func personBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "person", "PersonTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read person test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse person test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"person01", "person02", "person03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("TVMAZE_TEST_PERSON_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"TVMAZE_TEST_PERSON_ENTID": idmap,
		"TVMAZE_TEST_LIVE":      "FALSE",
		"TVMAZE_TEST_EXPLAIN":   "FALSE",
	})

	idmapResolved := core.ToMapAny(env["TVMAZE_TEST_PERSON_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["TVMAZE_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
			},
			extra,
		})
		client = sdk.NewTvmazeSDK(core.ToMapAny(mergedOpts))
	}

	live := env["TVMAZE_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["TVMAZE_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
